File: platform/macos/Storage/NativeRecentHistory.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Combine
import Foundation

// One app-wide presentation/operation owner. Connection success enqueues a
// bounded address only; no disk IO or await extends the connection operation.
@MainActor public final class NativeRecentHistory: ObservableObject {
  @Published public private(set) var endpoints: [String] = []
  @Published public private(set) var isBusy = false
  @Published public private(set) var error: NativeStorageError?
  @Published public private(set) var hasLoaded = false
  @Published public private(set) var importOfferDismissed = false
  @Published private var importEligible = false
  public var canImportHistory: Bool { canEdit && importEligible }
  public func dismissImportOffer() { importOfferDismissed = true }
  private let store: NativeProfileHistoryStore
  private var revision: UUID?
  private enum Request: Sendable { case read, record(String), remove(String, UUID?), clear(UUID?) }
  private var operation: Task<Void, Never>?
  private var pending: [String] = [] // Oldest first, coalesced to latest 20.
  private var refreshPending = false
  private var stopped = false
  public init(store: NativeProfileHistoryStore) { self.store = store }
  deinit { operation?.cancel() }
  public var canEdit: Bool { !stopped && !isBusy && hasLoaded && error == nil }
  public func reload() {
    guard !stopped else { return }
    guard operation == nil else { refreshPending = true; return }
    begin(.read)
  }
  public func recordSuccessful(_ endpoint: String) {
    guard !stopped else { return }
    guard !endpoint.isEmpty, endpoint.utf8.prefix(4097).count <= 4096, !endpoint.utf8.contains(0) else { error = .invalid; return }
    pending.removeAll(where: { $0 == endpoint }); pending.append(endpoint)
    pending = Array(pending.suffix(NativeProfileHistoryStore.historyCapacity))
    advance()
  }
  public func remove(_ endpoint: String) {
    guard canEdit, endpoints.contains(endpoint) else { return }
    begin(.remove(endpoint, revision))
  }
  public func clear() {
    guard canEdit, !endpoints.isEmpty else { return }
    begin(.clear(revision))
  }
  private func advance() {
    guard !stopped, operation == nil, error == nil else { return }
    if !pending.isEmpty { begin(.record(pending.removeFirst())) }
    else if refreshPending { refreshPending = false; begin(.read) }
  }
  private func begin(_ request: Request) {
    isBusy = true; error = nil
    let owner = store
    operation = Task { @MainActor [weak self] in
      do {
        let snapshot: NativeProfileHistorySnapshot
        switch request {
        case .read: snapshot = try await owner.read()
        case .record(let endpoint):
          let current = try await owner.read()
          snapshot = try await owner.recordRecent(endpoint, expected: current.revision)
        case .remove(let endpoint, let revision): snapshot = try await owner.removeRecent(endpoint, expected: revision)
        case .clear(let revision): snapshot = try await owner.clearHistory(expected: revision)
        }
        if let self, !self.stopped {
          self.revision = snapshot.revision; self.endpoints = snapshot.recentEndpoints; self.hasLoaded = true
          self.importEligible = snapshot.canImportHistory
        }
      } catch {
        if let self, !self.stopped {
          self.importEligible = false
          self.error = (error as? NativeStorageError) ?? .ioFailure
          self.refreshPending = false // No automatic retry loop after failure.
          if case .record(let endpoint) = request, !self.pending.contains(endpoint) {
            self.pending.insert(endpoint, at: 0)
            self.pending = Array(self.pending.suffix(NativeProfileHistoryStore.historyCapacity))
          }
          if case .read = request { self.hasLoaded = false; self.endpoints = []; self.revision = nil }
        }
      }
      self?.operation = nil; self?.isBusy = false; self?.advance()
    }
  }
  public func stop() {
    guard !stopped else { return }
    stopped = true; importEligible = false; pending.removeAll(); refreshPending = false; operation?.cancel()
  }
  public func close() async { stop(); await operation?.value }
}

File: platform/windows/TidyVNC.Native/Storage/NativeRecentHistory.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.Collections.Immutable;
using CommunityToolkit.Mvvm.ComponentModel;

namespace TidyVNC.Native.Storage;

/// <summary>
/// The app-wide recent connections owner (NativeRecentHistory.swift). A
/// successful connection only enqueues its destination; store IO runs one
/// operation at a time off the connection path. A failed record keeps the
/// destination pending (the latest 20) without an automatic retry loop.
/// UI thread only.
/// </summary>
public sealed partial class NativeRecentHistory(IUiDispatcher dispatcher, NativeProfileHistoryStore store) : ObservableObject
{
    private abstract record Request;
    private sealed record ReadRequest : Request;
    private sealed record RecordRequest(NativeConnectionDestination Destination) : Request;
    private sealed record RemoveRequest(NativeConnectionDestination Destination, Guid? Revision) : Request;
    private sealed record ClearRequest(Guid? Revision) : Request;

    private readonly List<NativeConnectionDestination> pending = []; // Oldest first.
    private Guid? revision;
    private Task? operation;
    private bool refreshPending, stopped, importEligible;

    [ObservableProperty] public partial ImmutableArray<NativeConnectionDestination> Connections { get; private set; } = [];
    [ObservableProperty] public partial bool IsBusy { get; private set; }
    [ObservableProperty] public partial NativeStorageError? Error { get; private set; }
    [ObservableProperty] public partial bool HasLoaded { get; private set; }
    [ObservableProperty] public partial bool ImportOfferDismissed { get; private set; }

    public bool CanEdit => !stopped && !IsBusy && HasLoaded && Error is null;
    /// <summary>History has never been written and holds nothing: the first-use import offer applies.</summary>
    public bool CanImportHistory => CanEdit && importEligible;

    public void DismissImportOffer() => ImportOfferDismissed = true;

    public void Reload()
    {
        UiThread.Require(dispatcher);
        if (stopped) return;
        if (operation is not null) { refreshPending = true; return; }
        Begin(new ReadRequest());
    }

    public void RecordSuccessful(NativeConnectionDestination destination)
    {
        UiThread.Require(dispatcher);
        if (stopped) return;
        if (destination.Endpoint.Length == 0 || destination.Endpoint.Contains('\0', StringComparison.Ordinal) ||
            System.Text.Encoding.UTF8.GetByteCount(destination.Endpoint) > 4096)
        {
            Error = NativeStorageError.Invalid;
            return;
        }
        pending.Remove(destination);
        pending.Add(destination);
        if (pending.Count > NativeProfileHistoryStore.HistoryCapacity) pending.RemoveRange(0, pending.Count - NativeProfileHistoryStore.HistoryCapacity);
        Advance();
    }

    public void Remove(NativeConnectionDestination destination)
    {
        UiThread.Require(dispatcher);
        if (!CanEdit || !Connections.Contains(destination)) return;
        Begin(new RemoveRequest(destination, revision));
    }

    public void Clear()
    {
        UiThread.Require(dispatcher);
        if (!CanEdit || Connections.IsEmpty) return;
        Begin(new ClearRequest(revision));
    }

    private void Advance()
    {
        if (stopped || operation is not null || Error is not null) return;
        if (pending.Count != 0)
        {
            var next = pending[0];
            pending.RemoveAt(0);
            Begin(new RecordRequest(next));
        }
        else if (refreshPending)
        {
            refreshPending = false;
            Begin(new ReadRequest());
        }
    }

    private void Begin(Request request)
    {
        IsBusy = true; Error = null;
        OnPropertyChanged(nameof(CanEdit));
        operation = RunAsync(request);
    }

    private async Task RunAsync(Request request)
    {
        try
        {
            var snapshot = request switch
            {
                RecordRequest record => await Record(record.Destination),
                RemoveRequest remove => await store.RemoveRecentAsync(remove.Destination, remove.Revision),
                ClearRequest clear => await store.ClearHistoryAsync(clear.Revision),
                _ => await store.ReadAsync(),
            };
            if (!stopped)
            {
                revision = snapshot.Revision;
                Connections = snapshot.Value.RecentConnections;
                HasLoaded = true;
                importEligible = snapshot.Value.CanImportHistory;
            }
        }
        catch (NativeStorageException error)
        {
            if (!stopped)
            {
                importEligible = false;
                Error = error.Error;
                refreshPending = false; // No automatic retry loop after a failure.
                if (request is RecordRequest record && !pending.Contains(record.Destination))
                {
                    pending.Insert(0, record.Destination);
                    if (pending.Count > NativeProfileHistoryStore.HistoryCapacity) pending.RemoveAt(pending.Count - 1);
                }
                if (request is ReadRequest) { HasLoaded = false; Connections = []; revision = null; }
            }
        }
        operation = null;
        IsBusy = false;
        OnPropertyChanged(nameof(CanEdit));
        OnPropertyChanged(nameof(CanImportHistory));
        Advance();
    }

    private async Task<NativeRecordSnapshot<NativeProfileHistory>> Record(NativeConnectionDestination destination)
    {
        var current = await store.ReadAsync();
        return await store.RecordRecentAsync(destination, current.Revision);
    }

    public void Stop()
    {
        if (stopped) return;
        stopped = true; importEligible = false; pending.Clear(); refreshPending = false;
    }

    public async Task CloseAsync()
    {
        Stop();
        if (operation is { } running) await running;
    }
}
